### ISSUE TITLE
Feature Stop for the TableEditor & the Main Editor

### ISSUE THREAD

**Issue by nathanlesage** [Zettlr/Zettlr#3657]
Looking at the issue tracker here, there are many issues demanding certain features for the table editor that work perfectly well for the regular editor. Also, more and more discussions start on our Discord, where people are asking why some functionality apparently does not seem to work.

With this issue, I would like to declare a temporary feature stop for the TableEditor and the main Markdown Editor.

## Why does the TableEditor not work as expected?

First things first, to clarify a few things about the TableEditor: the main reason why certain features such as autocorrect or snippet insertion do not work in the TableEditor is that, technically, the TableEditor and the main Markdown editor are two separate pieces of software.

Hence, everything we implement for the main editor we would have to duplicate for the TableEditor. This explains why snippets work in the main editor, but not while editing a table. This is neither a bug nor a feature, it is simply one of the many instances where I didn't have the time to replicate some functionality from the main editor inside the TableEditor.

Missing functionality includes, among other things, keyboard shortcuts for bold/italic/strikeout/etc. text, image pasting, any form of autocomplete (snippets, citations, file links, tags) and autocorrect.

I plan to implement all of that functionality eventually in the TableEditor, but not now.

## Why a feature stop?

The reason I'm declaring a feature stop is not that I want to abandon the TableEditor, but rather something else: One of the very next steps I have to do before we can begin to make Zettlr ready for 3.0 is to migrate from the current code editor (CodeMirror 5) to the new version (CodeMirror 6). Since the whole thing has changed a lot, this is not a matter of changing a few lines of code but to actually build up the complete editor from the ground up _again_.

This will require a lot of work, and I have the credible suspicion that this will also include large changes to the TableEditor itself. Therefore I want to first migrate to CodeMirror 6, then re-implement the TableEditor and see what I need to change, and finally, when that is done and the functionality has arrived at the same functionality we have now, lift the feature stop and continue to replicate all the editor features for the table editor (in case this is even necessary at that point).

## What does this mean?

The major effect this will have on the development of Zettlr is that, for now, I will completely ignore every feature request for both the main editor and the table editor. I will ignore these until I have opened & merged the Pull Request that performs the CodeMirror 5 -> CodeMirror 6 migration plus a few weeks/months of extensive testing.

After the migration is done and everything works as expected, I will resume implementing features.

**NOTE that this also includes a merge-stop for any PRs, so please do not attempt to implement these features yourselves; we are unable to merge them.** Open PRs that we already have (such as the codeblock copy button) are not affected by this, (a) because this would be unfair to the people and their work, and (b) so that I have the required code and can replicate this for the CM6 migration.

***

If you have any questions, please do not hesitate to ask them; I'll try to answer as extensively as possible.

**Comment by bxff** [comment]
This is great, I have been working with CM 6 for a while and maybe I can help you if you need any help.

> I plan to implement all of that functionality eventually in the TableEditor, but not now.

It seems like the editor inside of the cells of the table is not CM instance which is causing such problems. Though I am not 100% sure how one might go about adding a instance to the cell I have made a issue in the CM discussion form, which might help in the future: https://discuss.codemirror.net/t/nested-editors-kind-of/4654

Most likely you would want to work with the highlight package, with conjunction to the markdown parser extension architecture to get basic things formatting(styles) up and running:
* Here is the lezer markdown extension in play, which adds parsing for elements, documentation in the `readme`: https://github.com/lezer-parser/markdown/blob/main/src/extension.ts 
* Here is the highlight package in play, which adds highlighting to specified [tags added in the parser](https://github.com/lezer-parser/markdown/blob/1c43d7a141efd4e1c9c70d869e5e8a2beb378508/src/markdown.ts#L1833-L1856): https://github.com/voracious/ink-mde/blob/main/src/vendor/extensions/theme.ts

Of course for more complex stuff you would want to look into `PluginView`, `StateField`, `StateEffect`, and much more in both reference documentation, and examples written on codemirror.net. [Ink](https://github.com/voracious/ink-mde) is a CodeMirror 6 Markdown editor, which could be a good inspiration at times, which too I have referenced, and the [discussion form](https://discuss.codemirror.net/) is also a great place on getting insights and even for asking questions.

I hope this is helpful, it took me a while to understand the workings of CM 6, but its really worth it, once its clicks, its really seems any and everything is possible.

**Comment by MatveyM11** [comment]
In the current state Table Ediotr is entirely broken when it comes to actualy editing stuff, but just rendering of the tables that were made in other .md editing software seems to be fine.

**Comment by nathanlesage** [comment]
The table editor is absolutely not completely broken. It's a hassle, but not broken.

**Comment by MatveyM11** [comment]
@nathanlesage, in my case it's was saving changes in the cells in a one time out of ten. And that's pretty unusable.

**Comment by nathanlesage** [comment]
I can't believe 1/10. rather 9/10. do you have a screencast for this to demonstrate?

**Comment by MatveyM11** [comment]
@nathanlesage sure, look -
https://youtu.be/8M5sIHZXpUk

**Comment by nathanlesage** [comment]
Okay that is definitely weird, normally Zettlr should be writing the changes back to the document when you change cells.

Feel free to keep this weird behavior in mind and in case remind me during the beta phase for Zettlr 3 if the table editor works wrongly again! (I'm saying this because the table editor will need many changes to work with Codemirror 6; my personal hope is that it will be much more stable because of this, but I can't say for sure right now)

**Comment by dmint789** [comment]
I was wondering what the status of this is. I was going to open an issue with a bug that made it so that the table editor was not saving changes, but then I found this. Should I just avoid making tables with Zettlr until this rework is finished?

**Comment by nathanlesage** [comment]
The changes are all in the codebase now (but in the nightly). If your bug occurs in the nightly, please open an issue, if it occurs in 2.3, there's a high chance it has been fixed in the meantime.

**Comment by dmint789** [comment]
I am on 2.3 stable. Is it unsafe to use the current nightly?

**Comment by dmint789** [comment]
I have just tried it out and it works great, thanks a lot!

**Comment by nathanlesage** [comment]
That's great to hear! Yes, I think in general the nightly should be safe to use. There may be some odd bugs here and there but @Etienne-TR is doing an amazing job at finding them and filling up the issue tracker!

**Comment by Etienne-TR** [comment]
In nightly-20221128 I still have problems with saving tables :
1. When edit a cell, that doesn't turn the note to unsaved state
2. When select the all table, new contents in cells are erased

Is that working for you ?

**Comment by Etienne-TR** [comment]
I see there is a specific save button on editing a table. That works well if I click it !

However it seems to me very risky to lost table new content this way. I think it would be safer to turn the note to unsaved state on editing a cell (so that saving button gets useless and could be removed).

**Comment by Etienne-TR** [comment]
I disagree this is the best way. It seems better to me having to open a new window (or popup ?) than to loss content. Even more that table editor is not so easy for people that are not familiar and that would be very annoying to loss some work that was difficult to make.

I can't really encourage my co-workers editing tables in these conditions.

In release 2.3, table contents were saved like the others parts of a note. This is not possible in nightly anymore because of the changes made to be able to open multiple notes in the same window ?

**Comment by nathanlesage** [comment]
So you'd definitely prefer some overlay over the whole window so that you have to "accept changes" in some form in order to make it absolutely clear? If so, are there others against? It would indeed make it easier from a technical standpoint as well (because the separation of editing and displaying is clearer)
